pre-cache/parse_index.py

```diff
@@ -27,59 +26,76 @@ def parse_args():
-            bundle_dirs = [entry.name for entry in it if entry.is_dir()]
-        bundle_dir = sorted(
-            bundle_dirs, key=lambda dir_name: SequenceMatcher(
-                None, csv_name, dir_name).ratio())[-1]
-        bundles.append(os.path.join(
-            args.index_download_path, pkg_name, bundle_dir))
-    return bundles
+        if len(item) != 2:
+            print(f"operators record {item} is malformed, skipping...")
+            continue
+        package, channel = [i.strip() for i in item]
+        channels[package] = channel
+        packages.append(package)
+        print(f"will process package {package} channel {channel}")
+    # 2. Find the right channels for our packages and get the latest bundle
+    for item in objects:
+        if item.get("schema") == "olm.channel":
+            if item.get("package") in packages and item.get("name") == channels[item.get("package")]:
+                latest = item.get("entries")[-1].get("name")
+                bundles.append(latest)
+    # 3. extract related images from our bundles
+    for item in objects:
+        if item.get("schema") == "olm.bundle":
+            if item.get("name") in bundles and item.get("package") in packages:
+                images.extend([elem.get("image") for elem in item.get("relatedImages")])
+    return images
 
 
-def extract_images(bundles):
-    images = []
-    try:
-        for path in bundles:
-            with os.scandir(path) as it:
-                csv_file = [entry for entry in it if entry.name.endswith(
-                    '.clusterserviceversion.yaml')][-1]
-            with open(csv_file.path, 'r') as f:
-                csv = yaml.safe_load(f)
-            images.extend([im.get('image') for im in csv.get(
-                'spec').get('relatedImages')])
-    except Exception as e:
-        print(e, csv_file)
-        traceback.print_exc(file=sys.stdout)
-        raise Exception("Failed to extract related images", e)
-    finally:
-        return images
+def load_rendered_index():
+    with open(args.rendered_index.name, args.rendered_index.mode) as f:
+        data = f.read().lstrip()
+    # Rendered index is not a valid json, but a list
+    # of concatenated json blocks. Hence the raw decoder and the loop
+    decoder = json.JSONDecoder()
+    objects = []
+    while data:
+        obj, index = decoder.raw_decode(data)
+        objects.append(obj)
+        data = data[index:].lstrip()
+    return objects
 
 
 if __name__ == "__main__":
     try:
         args = parse_args()
-
-        bundles = extract_bundle_names(args)
-        images = extract_images(bundles)
+        data = load_rendered_index()
+        images = extract_images(args, data)
         with open(args.img_list_file.name, args.img_list_file.mode) as f:
             f.write('\n'.join(images))
             f.write('\n')
-        exit(0)
+        sys.exit(0)
     except Exception as e:
         print(e)
```

Comment:
I would use a logger here, rather than a print. Won't also be blocking the patch due to this but it's a good practice we may want to check for future reference.